### DISCUSSION TITLE
Technology Stack

### DISCUSSION THREAD

**Discussion by hischeapmoves** [xpollen8/jazzbutcher.com#51] [General]
Wherein the boffins discuss the technical implementation of this intended trove of material...

I read in the wiki that the plan is to store the site contents as JSON and then transform that into HTML. JSON is good, but not very human-readable or -editable. I propose a YAML + markdown format, which is very legible, easy to edit without special tools, and can be used by a variety of content renderers. I have some experience with one of these ([Jekyll](https://jekyllrb.com)) but there are many options.

Thoughts? Suggestions?

**Comment by wnerini** [comment]
> Wherein the boffins discuss the technical implementation of this intended trove of material...
> 
> I read in the wiki that the plan is to store the site contents as JSON and then transform that into HTML. JSON is good, but not very human-readable or -editable. I propose a YAML + markdown format, which is very legible, easy to edit without special tools, and can be used by a variety of content renderers. I have some experience with one of these ([Jekyll](https://jekyllrb.com)) but there are many options.
> 
> Thoughts? Suggestions?

So, thoughts... They're pretty stream-of-consciousness and all just "What If?" thoughts, but I'm a giant nerd and I'm going filter-free, so don't take anything as critical! Also, I'm going to be mixing in thoughts I've had about _very-long-term viability_ so watch out for that; I'm going to post a thread today with thoughts about viability 50 and even 100 years out (I've worked on two projects that included that as a requirement!)  

Let's look at David's roadmap, piece by piece...

**Media**
I think the solutions proposed, S3 short-term, with long-term dependency on archive.org is definitely the best approach we have for very long-term viability.

**Data (content and metadata/contextual data)**
My experience is that content and context should be stored together, so my thoughts will generally follow that philosophy.

_Data Storage_
I think the format that we store the data in is not as important as the fact that it's exportable or transformable into a open format. It should also be directly accessible to future maintainers, both as displayable, transformed output, as well as in its raw storage format. This gives future maintainers the flexibility to deal with format and deliver changes in what will be their modern format.  (Maybe HTML is deprecated!) 

_Data Generation_
Now, if the idea is that contributors will be creating data directly in our end-product structured storage format, I think that is going to be a rough road. 

I've some experiences with such an approach (users supplying data directly in a structured and transformable format, though with XML, not JSON (I work with JSON regularly these days )) and my experience is it's going to end up being a boffin-heavy situation.  I've seen non-techie folk will struggle with XML or YAML and we've got a lot of volunteers whose experience isn't technical.  The way to overcome that is with an editor (GitHub's is open source, I believe) that outputs the contribution.  I definitely agree that our storage format should be as human-readable as possible, to ensure that, in some mythical far-future an archeologist can regenerate the site even if the person looking at the data doesn't know anything about the web or html. 😃 

**Other Stuff**
_I think the only thing I would like us to discuss is looking at the data inputs and outputs with a generational not transformative mindset._ **But**, this is where I think we're getting into thoughts about very-long-term-viability, and are better in that thread.  But, in summary, I think we're definitely heading in the right direction with the roadmap, we can adjust course as a  group-consensus is decided (or, rather David is convinced or not 😸)

**Reply by aaroach** [reply]
As someone who is more on the front-end/configuration side of development, I like the idea of using an editor. I currently do edit TOML, HTML, and JSON in my day job, but definitely agree that something with a front-end is more efficient for content and asset management for many of us. I don't have experience with GitHub's editor to know if it's the answer (other than what I'm typing into right now). I did notice that when I tried to upload an .mp3, Git wouldn't allow it, so we'd need something that could take files of any type, I think. Perhaps the referenced open source editor is a different tool?

**Reply by wnerini** [reply]
I really just tossed out GitHub's content editor as a label, we'd have to decide what's right or code it when the time came.

**Reply by hischeapmoves** [reply]
It sounds like our priorities are aligned. I have my own trove of content that I hope future archeologists will be able to decode one day.

**Reply by xpollen8** [reply]
YAML - personally I detest it (spaces/tabs, twitchy as heck).
JSON is *practically* human-readable, but its huge benefit is the ecosystem around its production/consumption.  Browsers just *know* it.

And JSON is readily programmatically searchable - its isn't just a storage.

BUT.

Our goal is to select an approach that is *transformable* into the future, and JSON meets that need, as long as the *model* is accurate.

I've tried to get into JSON-LD, as it seems idea for what we're doing - has anyone played w/it?

@aaroach - Yes - there will always be some nice UX that will hide the details of how things are stored.

@wnerini - Absolutely - tying important metadata more closely to _a thing_ is very important. But the tape-trader method of _each gig is a zipfile_, while perfect in concept, is kinda hostile to the needs for: realtime delivery, searching, etc.

 Lacking a better framework, I've started to store simple .txt files along with media just to capture attribution.  Since we have to deal with objects (images, video, etc) that don't innately have ways to "tag" in a useful manner, we're stuck trying to enforce a workflow that accomplishes all of a) sticks w/that object b) is human reasable and is c) computer transformable.

Curated JSON that references external media has been my solution thus far.

@hischeapmoves JBC content is only one of my data preservation efforts.  (See [700 West Recording](https://700west.com), Family genelogical records w/artifacts, etc). I am astonished that the *World Of Museums* there doesn't seem to be a modern, available system for _maintaining physical/digital records_.  Such a system could assist here.



Longwinded, but the takeaway is: data storage and presentation are different things/we can solve the *editor* part distinct from the *data model*

**Reply by xpollen8** [reply]
> As someone who is more on the front-end/configuration side of development, I like the idea of using an editor. I currently do edit TOML, HTML, and JSON in my day job, but definitely agree that something with a front-end is more efficient for content and asset management for many of us. I don't have experience with GitHub's editor to know if it's the answer (other than what I'm typing into right now). I did notice that when I tried to upload an .mp3, Git wouldn't allow it, so we'd need something that could take files of any type, I think. Perhaps the referenced open source editor is a different tool?

Yo... I'm working on a dedicated media uploader for just this case.. Hang in there

**Reply by wnerini** [reply]
> xpollen8 34 minutes ago
> the tape-trader method of _each gig is a zipfile_, while perfect in concept, is kinda hostile to the needs for: realtime delivery, searching, etc.

As an old school scenster, I know the 360k/1.4m floppy-disk limit that spawned that practice, and it's a pet peeve of mine, and people still us it in the wild: drives me nuts.  I was really referring more to what you mention below, typing the media file and metadata together in a model. 

> I've tried to get into JSON-LD, as it seems idea for what we're doing - has anyone played w/it?
> 
I dunno, seems like way-big overkill, for modeling media and page-content data & location. I'm convincible though.

> Lacking a better framework, I've started to store simple .txt files along with media just to capture attribution. Since we have to deal with objects (images, video, etc) that don't innately have ways to "tag" in a useful manner, we're stuck trying to enforce a workflow that accomplishes all of a) sticks w/that object b) is human reasable and is c) computer transformable.
> 
All modern media formats have meta-data. It may not be sufficiently rich for our purposes (for certain in some cases), but we should strive to ensure we push the rich information downstream into the media files. 

> Curated JSON that references external media has been my solution thus far.
> 
That's really the thing I was speaking about in my "content and context should be stored together," Not necessarily on top of each other.  Simplistically: 
```
shows[0] = {
  title: "",
  date: "",
  location: "",
  title: "",
  uri: ""
  etc: ""
}
```

> Longwinded, but the takeaway is: data storage and presentation are different things/we can solve the _editor_ part distinct from the _data model_

100% in agreement.

**Reply by xpollen8** [reply]
> All modern media formats have meta-data. It may not be sufficiently rich for our purposes (for certain in some cases), but we should strive to ensure we push the rich information downstream into the media files.
> 
> > Curated JSON that references external media has been my solution thus far.
> 
> That's really the thing I was speaking about in my "content and context should be stored together," Not necessarily on top of each other. Simplistically:
> 
> ```
> shows[0] = {
>   title: "",
>   date: "",
>   location: "",
>   title: "",
>   uri: ""
>   etc: ""
> }
> ```

Thing is, the model is actually a very very complex many:many relationship.

Performer has [ Gigs, Releases, Photos, Ancedotes, ... ]
Gig has [ Performers, Attendees, Recording, Photo, Setlist, Songlist, Ticket, Event Website, Reviews... ]
Attendee has [ Gigs, Recordings, Ancedotes... ]
Release has [ Demo, Reviews, Performers... ]
Photo has [ Photographer, Ancedote, Source... ]
and on and on

Additionally, some of this information is backed by *Real World* physical artifacts. Pat's notebooks (documenting song history, the source of MANY of the site's gig dates).  Should each gig denote that its information came from Pat's notebook versus setlists.fm... Absolutely.  Is it overkill. Absolutely.

On top of this, we may soon-ish need to ingest a flood of new material -  the contents of Pat's home. Lots of original source material.  To be digitized and curated.

Big tasks up ahead which we need to plan for.  We gotta get the model right to capture.

**Reply by wnerini** [reply]
> Thing is, the model is actually a very very complex many:many relationship.
> 
> Performer has [ Gigs, Releases, Photos, Ancedotes, ... ] Gig has [ Performers, Attendees, Recording, Photo, Setlist, Songlist, Ticket, Event Website, Reviews... ] Attendee has [ Gigs, Recordings, Ancedotes... ] Release has [ Demo, Reviews, Performers... ] Photo has [ Photographer, Ancedote, Source... ] and on and on
> 
Yeah, it's gonna be complex, and maybe even get more, as we find new things that need storing. 

> Additionally, some of this information is backed by _Real World_ physical artifacts. Pat's notebooks (documenting song history, the source of MANY of the site's gig dates). Should each gig denote that its information came from Pat's notebook versus setlists.fm... Absolutely. Is it overkill. Absolutely.
> 
This is a preservation effort, so my opinion is definitely save it all. Who knows what future cultural historians might need!

> On top of this, we may soon-ish need to ingest a flood of new material - the contents of Pat's home. Lots of original source material. To be digitized and curated.
> 
> Big tasks up ahead which we need to plan for. We gotta get the model right to capture.

Agreed: the model is the thing.

**Comment by xpollen8** [comment]
*How Things Are Currently Implemented*

JBC com is _mostly_ delivered via mysql, where gig and gig_personnel, press data is stored.  *That* data is easily transformed into other forms.

Much of the rest is pseudo structured data: Albums, tracklistings, people, etc - Are grouped together in the *HTDB* language code.  This data is _reasonably_ transformable into something else, but unfortunately, that task'll be on me.  *HTDB* takes some getting used to.

That said.  I've already taken a whack at making the gig/related data accessible as an API: [data.jazzbutcher.com](https://github.com/xpollen8/data.jazzbutcher.com/blob/main/pages/api/%5B%5B...path%5D%5D.js) 

Examples:
   [Here are all the gigs](https://data.jazzbutcher.com/api/gigs)
   [Here are all the interviews](https://data.jazzbutcher.com/api/press)
   [Here are all the gigs w/audio](https://data.jazzbutcher.com/api/gigs_with_audio)
   [How many times "Scarlett" was played live](https://data.jazzbutcher.com/api/live_performances_by_song/scarlett)
   [Max Eider interviews](https://data.jazzbutcher.com/api/interviews_by_person/Max_Eider)
   
data.jazzbutcher.com is just a proof of concept at the moment, but you can imagine that with *a full suite* of APIs, once could easily re-write JBC com around them.

**Reply by wnerini** [reply]
I think the models (and the API!) look fine, only small things I'll eventually see, if past experience is anything to go by.  I'm going to go out on a limb and say that your API signatures are 100% correct! 😉 

I'm definitely trying to think of this less as a website project, than as a project to create an archival data-set not tied to any particular technology stack, from which the future (and us in the not-so-future) could generate any sort of displayable or printable content about the JBC. 😄 

_I truly think, with the right approach, it's possible to make this a project that can publish something that can last a very, very long time, even after we're all dust._  What that approach is, I'm still undecided about.

**Reply by xpollen8** [reply]
I guess the question is: _Is the archival format the *source*, or one of the *data views*_?

Does one derive API from the archival, or is the archival a snapshot of API output?

I think archival is the *end product*: make a change to the dataset, some automated process wraps it up and pushes to archive.org

**Reply by wnerini** [reply]
Aha! Now we're getting into what I'd think of long-term solutioning. I'll address that in the thread I was talking about earlier, but for now, I don't think your current conceptions are un-workable or anything, they're best-practices and all, and they're fleshed out so my opinion is we hew closely to them.

However, here's some teaser answers to what we can deal with in the "long-scale" chat I'd like to have.

> I guess the question is: _Is the archival format the _source_, or one of the _data views__?
> 
The data is the archive, and the archive is the data _(ooh, how Zen of me! 😉 )_

> Does one derive API from the archival, or is the archival a snapshot of API output?
> 
So, I don't even see an API, in the future.... _(Hmm!)_

> I think archival is the _end product_: make a change to the dataset, some automated process wraps it up and pushes to archive.org
>
I think this is pretty close to what I'm thinking. It's at least walking down the same road! 😄

**Reply by xpollen8** [reply]
@wnerini (and anyone else) Are you up for continuing this conversation?

I could use a perspective check.

The past 6 weeks *solid* have been me porting JBC com into a data-backed React app. 10 hours a day, every day.  My family rightly hates me.

Some things which were not entirely in the database now are: `lyrics`, `press`.  I'm in the middle of doing the same for `conspirators` and last will be the `releases`.  I knocked off the `gigs` first.  What a bastard that was! *FIshy Mansions* looks real nice now. Usable on mobile even.

Since V1 needs to continue to work while this process occurs, my approach has been to have a data schema that both V1 and V2 understand.  Which means I've been incrementally changing the schema under V1 and having to update the V1 code in tandem with developing V2.  The goal is then to optimize and consolidate the schemas after V2 is switched live.  

It is only *then* that we can truly discuss *What Is The Right Forevertime Schema To Store The Data* topic.

Of course, I've been performing many V2 side-quests; incorporating all the "pre-JBC" project information and archiving as much `press` content as possible.  And @Bigquestionsinc  keeps finding me stuff to include.  Not complaining!

All the V2 data comes from API requests to data.jazzbutcher.com.  V2 no longer makes direct database calls - it only queries the API.

Oh yeah,  V2 will have an `admin` portal.  Auth is already working. I anticipate that the admin section will allow for direct editing of the datasets and probably an S3 file navigator.

I'd just like to chat about this stuff with someone.  It feels like I'm off in the wilderness and it is exhausting.

I'm open to a livestream show-and-tell.

Could someone with FB/Meta access please query the JBC forums to see if there are programmers (nextjs/React) or web/CSS specialists who might like to join us here?  Otherwise, y'all are going to be stuck with my website design sensibilities for another 30 years.

**Reply by wnerini** [reply]
> Could someone with FB/Meta access please query the JBC forums to see if there are programmers (nextjs/React) or web/CSS specialists who might like to join us here? Otherwise, y'all are going to be stuck with my website design sensibilities for another 30 years.

That would be me. 3 decades web app experience. Not the react, it's not shown itself as useful in the enterprise apps I manage to encourage me to put time into, but what's another stack, among dozens I already know. But  CSS/layout, that's cross-framework. I can at least give a couple of relatively experience-based opinions.

I'll post something in the FB forums today, as well. More perspectives, more help.  

And just set us up a time, and we can have a discussion.  I'd love to talk about how we can future-proof the hell out of the website, for when we're all dead and gone. :)

**Reply by hischeapmoves** [reply]
I'm available as well... no React experience (work uses Vue.js), but I can certainly help with HTML / CSS / UI

**Reply by xpollen8** [reply]
Ian!  I didn't realize that you were ink19!  I'm listening to your 2006 interview. Our origin story is almost identical, except I'm 3 year ahead of you in the time line, in a different part of the country. and I went off into startup land whle you managed to make your hobby pay. Congrats

@wnerini , @hischeapmoves (and anyone else) ...

I *just* added instructions on how to run V2 yourself:  https://github.com/xpollen8/jazzbutcher.com/blob/main/README.md

Spin it up and see how it goes.

I'd love to have a realtime chat with tech types just to shoot the shit and go over where I'm headed and where I could use help.

**Reply by hischeapmoves** [reply]
Wow, flattered to be recognized. Strangely, these days I live in startup land, my day job is with a VC fund.

I am traveling for the next couple of weeks, so my access is irregular, but I have my laptop and I'll see about getting the site spun up locally in the meantime.

A regular tech chat session would be quite useful!

—Ian

**Reply by xpollen8** [reply]
If you do get it running, be sure to `git pull` *frequently*.  I'm averging tens of checkins each day.

Currently, I'm doing two things in parallel:

* porting V1 features
* dragging V1 data into an interim schema that allow me to move V2 to a better world while minimally impacting V1
   most of this work is very weedsy, and pretty much on me, since I knocked the ball over there to begin with.

That second point is forcing me to think hard about the *true data model*.

For example: right now some things are well described: gig songs - most every bit of associated metadata is already in the schema.

Weird corner cases, which I'm finding are not a good fit: one-off side-project records, or the "Masters of Budvar" shows where Pat didn't actually play.  Both of those are currently shoehorned into the "Gig" model with special flags so that the rendering can take their "specialness" into account.  Oh, and those Eg Demo tapes.  Same deal.  And Wilson.  And the fact that I've always put Interviews into that same bucket.

I'm now at the point where the metadata for each of those types of things vary quite a bit.  So the old "gig+song" data model is not one to keep going forward.

The replacement .. I haven't yet zeroed in on.

But for now, I have a few frequently evolving interim database tables to capture all those weird bits. With the understanding that once we all have a sit back and look - we can find what shape the master schema should be.

-----

Longwinded, but this is to say: some of the code will be utterly perplexing simply due to having to account for the current messy schema.

But the overall arch of V2 is pretty solid and easy to navigate, I think.

-----

@hischeapmoves Here's a quck question: for ink19, when you had obvious link-outs embedded inside of articles, how did you maintain those?

In V1, I would use macros expansions to accomplish the goal:

```
#define pat <a href="...." class="..">Pat Fish<a>
#define kittens <a href="...." class="..">Love Kittens</a>
```

and then use it like:
```
In this interview, ${pat} discusses ${kittens}
```

This works just fine.. for V1.  The problem is that those V1 expansions are now already-expanded when seen in V2:

```
In this interview, <a href"...." class="..">Pat Fish<a> discusses <a href="...." class="..">Love Kittens</a>
```

Not ideal, since V2 may want to handle expansions differently, and now the original `pat` and `kittens` lookup variables are absent.

-----

My options are:

* disable V1 expansions (very difficult) for a one-time scrape of everything so that `${pat}` ends up in V2
* have V1 embed the original lookup variable in HTML comment (`<!-- expanded_from:pat -->`) near the expansion, and have V2 parse it out (doable)
* have V2 completely strip all HTML that surrounds the expanded `pat`, and then use a render-time `AutoLink` feature that expands `Pat Fish` into whatever V2 wants `Pat Fish` to render as.  I've used this approach elsewhere. Mostly works. Hard to get right. Recursion. Voodoo.

None of these are ideal.

Curious whether ink19 used any sort of {{pat}} like rendertime lookup, or whether you just said *screw it*, and embedded links directly in articles.  Or something else.
